src/Npgsql/NpgsqlDataAdapter.cs

```diff
@@ -137,6 +140,74 @@ protected override void OnRowUpdating(RowUpdatingEventArgs value)
             get => (NpgsqlCommand?)base.InsertCommand;
             set => base.InsertCommand = value;
         }
+
+        internal Task<int> Fill(DataTable dataTable, bool async, CancellationToken cancellationToken = default)
+        {
+            if (async)
+                return FillAsync(dataTable, cancellationToken);
+            else
+                return Task.FromResult(Fill(dataTable));
+        }
+
+        // Temporary implementation, waiting for official support in System.Data via https://github.com/dotnet/runtime/issues/22109
+        async Task<int> FillAsync(DataTable dataTable, CancellationToken cancellationToken = default)
+        {
+            var command = SelectCommand;
+            var activeConnection = command?.Connection ?? throw new InvalidOperationException("Connection required");
+            var originalState = ConnectionState.Closed;
+
+            try
+            {
+                originalState = activeConnection.State;
+                if (ConnectionState.Closed == originalState)
+                    await activeConnection.OpenAsync(cancellationToken);
+
+                var behavior = CommandBehavior.SequentialAccess;
+                using var dataReader = await command.ExecuteReaderAsync(behavior, cancellationToken);
+
+                return await FillAsync(dataTable, dataReader);
```

Comment:
Missing cancellation token. If you rebase on latest main this becomes a build error.